src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
+        // Loop Though all the edges
+        let ct1 = cmp::max(input.all_edge_cost.borrow().cost_move, input.all_edge_cost.borrow().cost_copy);
+
+        let ct = cmp::max(ct1, input.all_edge_cost.borrow().cost_glue); // ct the max value of Move , copy and glue
+        let all_edges = input.list_edges.borrow().clone();
+        let mut edges_to_be_pruned: Vec<Edge> = Vec::new();
+        let mut edges_still_left: Vec<Edge> = Vec::new();
+        // Initial Pruning Guess
+        for edge in &all_edges {
+            // edge variable will be E1 --> (m,n)
+            // Get All The edges for E2 --> where (m,n1)
+            // Where the from_node_id is the same but to_id is different
+            // Get All The edges for E3 ---> where (m1,n)
+            // Where the to_node_id is the same but to from_id is different
+            let mut upper_bound_e2_and_e3 = 0;
+
+            // E2 = (m,n1) &&   E3 = (m1,n)
+
+            upper_bound_e2_and_e3 = input
+                .clone()
+                .upper_bound_edge(edge.clone(), all_edges.clone());
+            //
+
+            //
+            let mut lower_bound_e1 = 0;
+            lower_bound_e1 = input
+                .clone()
+                .lower_bound_edge(edge.clone(), all_edges.clone());
+
+            if lower_bound_e1 >= upper_bound_e2_and_e3 + 2 * ct {
+                edges_to_be_pruned.push(edge.clone());
+            } else {
+                edges_still_left.push(edge.clone());
+            }
+        }
+
+        let remaining_edges = edges_still_left.clone();
+        // Remove the update edges if the number of descendants are not the same.
+        let mut edges_after_pruning_update: Vec<Edge> = Vec::new();
+        for e1_update in remaining_edges.clone() {
+            if e1_update.edge_type == EdgeType::UPDATE {
+                let m = e1_update
+                    .from_node
+                    .descendants(&input.from_arena.borrow())
+                    .collect::<Vec<NodeId<FromNodeId>>>()
+                    .len();
+                let n = e1_update
+                    .to_node
+                    .descendants(&input.to_arena.borrow())
+                    .collect::<Vec<NodeId<ToNodeId>>>()
+                    .len();
+                // if the number of descendants are the same then ok to update
+                // Otherwise you may need to insert nodes and delete etc so update won't make any sense.
+                if m == n {
+                    edges_after_pruning_update.push(e1_update.clone());
+                }
+            } else {
+                edges_after_pruning_update.push(e1_update.clone());
+            }
+        }
+
+        edges_still_left = edges_after_pruning_update;
+
+        // The Root should not move just update.
+        // If it's root it should not have move, then it should have insert as well
+        //          Which doesn't make any sense
+        // It could have glue, copy
```

Comment:
I'm not sure I follow this comment. What happens if the two root nodes have different types and different labels?